File: MasiWasiMovil/app/src/main/java/viewmodels/HomeViewModel.java
package viewmodels;

import android.util.Log; // Importante para depurar
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import java.util.ArrayList;
import java.util.List;
import models.Mascota;

public class HomeViewModel extends ViewModel {

    private final MutableLiveData<List<Mascota>> mascotasLiveData = new MutableLiveData<>();
    private final List<Mascota> mascotasOriginal = new ArrayList<>();
    private final FirebaseFirestore db = FirebaseFirestore.getInstance();

    public HomeViewModel() {
        cargarDatosDesdeFirestore();
    }

    private void cargarDatosDesdeFirestore() {
        // Usamos SnapshotListener para tiempo real
        db.collection("mascotas")
                .addSnapshotListener((value, error) -> {
                    if (error != null) {
                        Log.e("FirestoreError", "Error al cargar mascotas: " + error.getMessage());
                        return;
                    }

                    if (value != null) {
                        mascotasOriginal.clear();
                        for (DocumentSnapshot doc : value.getDocuments()) {
                            Mascota m = doc.toObject(Mascota.class);
                            if (m != null) {
                                m.setId(doc.getId());
                                mascotasOriginal.add(m);
                            }
                        }
                        Log.d("FirestoreData", "Mascotas cargadas: " + mascotasOriginal.size());
                        // Usamos postValue para asegurar que se actualice la UI incluso desde hilos secundarios
                        mascotasLiveData.postValue(new ArrayList<>(mascotasOriginal));
                    }
                });
    }

    public LiveData<List<Mascota>> getMascotas() {
        return mascotasLiveData;
    }

    public void filterBy(String categoria) {
        if (categoria.equalsIgnoreCase("Todos")) {
            mascotasLiveData.setValue(new ArrayList<>(mascotasOriginal));
        } else {
            List<Mascota> filtradas = new ArrayList<>();
            for (Mascota m : mascotasOriginal) {
                // CORRECCIÓN: Evitar NullPointerException si la categoría viene nula de Firestore
                if (m.getCategoria() != null && m.getCategoria().trim().equalsIgnoreCase(categoria.trim())) {
                    filtradas.add(m);
                }
            }
            mascotasLiveData.setValue(filtradas);
        }
    }
}

File: movil2/mozita/app/src/main/java/viewmodels/HomeViewModel.kt
package viewmodels

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import models.Mascota
import repository.MascotaRepository

// Cambiamos ViewModel() por AndroidViewModel(application)
class HomeViewModel(application: Application) : AndroidViewModel(application) {

    private val mascotasLiveData = MutableLiveData<List<Mascota>>()
    private val mascotasOriginal: MutableList<Mascota> = ArrayList()
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()
    // Inicializamos nuestro Repositorio pasándole el application (que actúa como contexto)
    private val mascotaRepository = MascotaRepository(application)

    init {
        //Llamamos primero a nuestra base de datos local
         cargarDatosLocales()


         cargarDatosDesdeFirestore()
    }

    // fun Leer desde SQLite
    private fun cargarDatosLocales() {
        try {
            // Le pedimos al repositorio que traiga la lista de mascotas
            val listaLocal = mascotaRepository.obtenerMascotas()

            mascotasOriginal.clear()
            mascotasOriginal.addAll(listaLocal)

            Log.d("SQLiteData", "Mascotas cargadas desde SQLite: ${mascotasOriginal.size}")

            // Actualizamos la UI para que el RecyclerView las pinte
            mascotasLiveData.postValue(ArrayList(mascotasOriginal))

        } catch (e: Exception) {
            Log.e("SQLiteError", "Error al cargar mascotas locales: ${e.message}")
        }
    }

    private fun cargarDatosDesdeFirestore() {

        // Escuchamos los cambios en Firebase en tiempo real
        db.collection("mascotas")
            .addSnapshotListener { value, error ->

                if (error != null) {
                    Log.e("FirestoreError", "Error al cargar mascotas: ${error.message}")
                    return@addSnapshotListener
                }

                if (value != null) {
                    // SINCRONIZACIÓN
                    for (doc: DocumentSnapshot in value.documents) {
                        val m = doc.toObject(Mascota::class.java)
                        if (m != null) {
                            m.setId(doc.id)

                            // Guardamos/Actualizamos la mascota de la nube en nuestro celular
                            mascotaRepository.insertar(m)
                        }
                    }

                    Log.d("Sync", "Sincronización con Firestore completada")

                    //  Volvemos a leer la base local para que la interfaz muestre los nuevos datos
                    cargarDatosLocales()
                }
            }
    }

    fun getMascotas(): LiveData<List<Mascota>> {
        return mascotasLiveData
    }

    fun filterBy(categoria: String) {

        if (categoria.equals("Todos", ignoreCase = true)) {
            mascotasLiveData.value = ArrayList(mascotasOriginal)

        } else {
            val filtradas: MutableList<Mascota> = ArrayList()

            for (m in mascotasOriginal) {

                // Evitar NullPointerException
                if (m.getCategoria() != null &&
                    m.getCategoria()!!.trim().equals(categoria.trim(), ignoreCase = true)
                ) {
                    filtradas.add(m)
                }
            }

            mascotasLiveData.value = filtradas
        }
    }
}